Fix #1139 - Log study enrollment events via Telemetry Events

mythmon opened the pull request:
This add documentation and implementation for enrollment events when studies start and stop. This includes both preferences and add-on studies.

@georgf Can you review this? The changes in `recipe-client-addon/` should be treated like Firefox code, and will eventually make its way into mozilla-central.

georgf commented on recipe-client-addon/lib/TelemetryEvents.jsm:
```diff
@@ -0,0 +1,61 @@
+      },
+    });
+
+    Services.telemetry.setEventRecordingEnabled(TELEMETRY_CATEGORY, true);
```
You don't need this for events that are registered at runtime through `registerEvents()`.
This is only needed for events shipping in Firefox code (to avoid collecting all the time).

georgf commented on recipe-client-addon/lib/TelemetryEvents.jsm:
```diff
@@ -0,0 +1,61 @@
+    isStarted = true;
+
+    CleanupManager.addCleanupHandler(() => {
+      Services.telemetry.setEventRecordingEnabled(TELEMETRY_CATEGORY, false);
```
Ditto, this is not needed.

georgf commented on recipe-client-addon/lib/TelemetryEvents.jsm:
```diff
@@ -0,0 +1,61 @@
+    if (isStarted) {
+      Services.telemetry.recordEvent(TELEMETRY_CATEGORY, method, object, value, extra);
+    } else {
+      log.warn(
```
Telemetry already logs errors when you try to record events that are not registered.
Maybe that is sufficient?

mythmon commented:
Thanks for the review, I've unassigned you, and will remove the unneeded parts you pointed out.

@gijsk can you review the rest of the code? Most of it should be pretty straight forward. I'm also working with PI to do manual QA of this feature, to make sure it matches the spec laid out in the docs.

gijsk commented on docs/user/data_collection.rst:
```diff
@@ -83,3 +83,106 @@ reported under the key ``shield-recipe-client/recipe/<recipe id>``:
+           was disabled, or the user no longer matches the recipe's
+           filter.
+         * ``"user-preference-changed"``: The study preference was
+           changed on the user branch. This ocould mean the user
```
Nit: `ocould`

gijsk commented on docs/user/data_collection.rst:
```diff
@@ -83,3 +83,106 @@ reported under the key ``shield-recipe-client/recipe/<recipe id>``:
+      addonId
+         The add-on's ID (example: ``"feature-study@shield.mozilla.com"``).
+      addonVersion
+         The add-on's ID (example: ``"1.2.3"``).
```
Nit: version

gijsk commented on docs/user/data_collection.rst:
```diff
@@ -83,3 +83,106 @@ reported under the key ``shield-recipe-client/recipe/<recipe id>``:
+      addonId
+         The add-on's ID (example: ``"feature-study@shield.mozilla.com"``).
+      addonVersion
+         The add-on's ID (example: ``"1.2.3"``).
```
Same here.

gijsk commented on docs/user/data_collection.rst:
```diff
@@ -83,3 +83,106 @@ reported under the key ``shield-recipe-client/recipe/<recipe id>``:
+      addonVersion
+         The add-on's ID (example: ``"1.2.3"``).
+      reason
+         The reason for unenrollment. Possible values are:
```
From the code, looks like this could also be `"unknown"` ? But that value is not listed here...

gijsk commented on recipe-client-addon/lib/PreferenceExperiments.jsm:
```diff
@@ -377,8 +382,11 @@ this.PreferenceExperiments = {
+            resetValue: false,
+            reason: "user-preference-changed",
+          })
+          .catch(Cu.reportError);
```
I would put this on the previous line, but ymmv.

gijsk commented on docs/user/data_collection.rst:
```diff
@@ -83,3 +83,106 @@ reported under the key ``shield-recipe-client/recipe/<recipe id>``:
+   extra
+      resetValue
+         The string ``"true"`` if the preference was set back to its
+         original value, ``"false"`` if it was left as its current
```
Out of curiosity, why strings that contain booleans? Why not just booleans?

Also, in the epitome of nitpicks, the name of this property makes me think its value is the value the preference was reset to, not whether or not it was reset. Dunno if there are obvious better choices. `didResetValue` or something?

mythmon replied:
According to [the docs](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/events.html#recordevent), API only allows "string keys & values", so booleans aren't an option.

I think you're right about the name of `resetValue`. There isn't a lot of outside-of-the-code history to this name, so now is a good time to change it.

gijsk commented on recipe-client-addon/lib/ShieldPreferences.jsm:
```diff
@@ -73,22 +73,24 @@ this.ShieldPreferences = {
     switch (prefName) {
       // If the FHR pref changes, set the opt-out-study pref to the value it is changing to.
-      case FHR_UPLOAD_ENABLED_PREF:
+      case FHR_UPLOAD_ENABLED_PREF: {
```
Huh, are the braces a new eslint thing? I don't mind either way, just surprised...

mythmon replied:
They aren't an eslint thing. I personally prefer them because it creates a scope for `let` and `const`. It doesn't apply here, but doing it consistently makes sense to me.

gijsk commented on recipe-client-addon/lib/TelemetryEvents.jsm:
```diff
@@ -0,0 +1,61 @@
+
+  sendEvent(method, object, value, extra) {
+    if (isStarted) {
+      Services.telemetry.recordEvent(TELEMETRY_CATEGORY, method, object, value, extra);
```
@georgf when, if ever, does `recordEvent` throw?

gijsk commented on recipe-client-addon/test/browser/browser_AddonStudies.js:
```diff
@@ -218,7 +226,7 @@ decorate_task(
   async function testStopNoStudy() {
     await Assert.rejects(
-      AddonStudies.stop("does-not-exist"),
+      AddonStudies.stop("does-not-exist", "test"),
```
Out of curiosity, what breaks if you don't provide a reason here? I thought there was a default value for the argument?

gijsk replied:
To be clear, it makes sense to want to test the propagation of the reason, but that doesn't happen here - you've just updated all the existing consumers and I'm not entirely sure why.

mythmon replied:
To help with my development, I had made all the functions that take `reason` require it, and throw if it was not passed. Then I made all the tests pass, including new tests that included reason propagation. After I got it working, I removed the error throwing and provided a default value to make this code less likely to throw errors.

Thinking about it more, I should have either left the error reporting in, or gone through and removed the no-op reasons in the tests. Do you have opinions about which would be better?

gijsk replied:
I don't really feel all that strongly either way. I think not throwing (and so removing the no-op reasons) might be a slightly safer bet given the few cases in the last 6 months or so where we had issues on startup. It might be nice to have a test that ensures that we end up with the default "unknown" value when no reason is passed.

gijsk commented on recipe-client-addon/test/browser/browser_PreferenceExperiments.js:
```diff
@@ -611,19 +647,36 @@ decorate_task(
+        experimentType: "exp",
+        branch: "branch",
+      }],
+      "start should send the correct telemetry event"
```
I like to have unique failure messages so I can quickly find out where things failed, but it's up to you.

mythmon commented:
I rebased this on master to fix the merge conflicts, and incorporated the feedback from @georgf and @gijsk. Apologies that I wasn't able to make it a second commit for a good interdiff, the rebase was not kind to workflow.

@gijsk Can you give this another pass and, if it is ready, tell bors r+?

I don't know if this is needed, or if you already have access, but it can't hurt:

bors delegate= @gijsk

gijsk commented on recipe-client-addon/lib/PreferenceExperiments.jsm:
```diff
@@ -454,8 +461,11 @@ this.PreferenceExperiments = {
    *   experiment has already expired.
    */
-  async stop(experimentName, resetValue = true) {
+  async stop(experimentName, {resetValue = true, reason = "unknown"} = {}) {
```
Nit: please update the docs.

gijsk commented on recipe-client-addon/lib/AddonStudies.jsm:
```diff
@@ -304,17 +325,17 @@ this.AddonStudies = {
    *   If the study is already inactive.
    */
-  async stop(recipeId) {
+  async stop(recipeId, reason) {
```
Nit: please update the docs.

gijsk commented on recipe-client-addon/lib/AddonStudies.jsm:
```diff
@@ -93,11 +92,21 @@ function getStore(db) {
  * @param {Study} study
  */
-async function markAsEnded(db, study) {
+async function markAsEnded(db, study, reason = "unknown") {
```
Docs here too

gijsk commented on recipe-client-addon/test/browser/browser_PreferenceExperiments.js:
```diff
@@ -752,8 +805,8 @@ decorate_task(
     ok(stopStub.calledWith("test"), "Experiment is stopped because value changed");
-    is(Preferences.get("fake.preference"), "changed value", "Preference value was not changed");
-  }
+    ok(Preferences.get("fake.preference"), "changed value", "Preference value was not changed");
```
Err, this is a strange change. `ok` only takes 1 test argument, and its value should be truthy for it to pass. Is this just a mistake? :-)

mythmon commented:
> as I don't want to accidentally get it to merge this before these last tiny bits of feedback are addressed.

Good call. Bors actually does the merging and doesn't understand r+wc. So had you told it r+, it would have done exactly what you feared. I've updates the docs, so I think this is good to go now.

bors r= @gijsk, @georgf
